src/Native/System.IO.Compression.Native/pal_zlib.cpp

```diff
@@ -0,0 +1,184 @@
+
+    stream->internalState = zStream;
+}
+
+/*
+Frees any memory on the PAL_ZStream that was created by Init.
+*/
+static void End(PAL_ZStream* stream)
+{
+    z_stream* zStream = reinterpret_cast<z_stream*>(stream->internalState);
+
+    delete zStream;
+    stream->internalState = nullptr;
+}
+
+/*
+Transfers the output values from the underlying z_stream to the PAL_ZStream.
+*/
+static void TransferState(z_stream* from, PAL_ZStream* to)
+{
+    to->nextIn = from->next_in;
+    to->availIn = from->avail_in;
+
+    to->nextOut = from->next_out;
+    to->availOut = from->avail_out;
+
+    to->msg = from->msg;
+}
+
+/*
+Transfers the input values from the PAL_ZStream to the underlying z_stream object.
+*/
+static void TransferState(PAL_ZStream* from, z_stream* to)
+{
+    to->next_in = from->nextIn;
+    to->avail_in = from->availIn;
+
+    to->next_out = from->nextOut;
+    to->avail_out = from->availOut;
+}
+
+/*
+Gets the current z_stream object for the specified PAL_ZStream.
+
+This ensures any inputs are transferred from the PAL_ZStream to the underlying z_stream,
+since the current values are always needed.
+*/
+static z_stream* GetCurrentZStream(PAL_ZStream* stream)
+{
+    z_stream* zStream = reinterpret_cast<z_stream*>(stream->internalState);
+    TransferState(stream, zStream);
+    return zStream;
+}
+
+extern "C" int32_t IsZLibAvailable()
+{
+    return true;
+}
+
+extern "C" int32_t DeflateInit2_(PAL_ZStream* stream, int32_t level, int32_t method, int32_t windowBits, int32_t memLevel, int32_t strategy)
+{
+    Init(stream);
+
+    z_stream* zStream = GetCurrentZStream(stream);
+
+    int32_t result = deflateInit2(zStream, level, method, windowBits, memLevel, strategy);
+    TransferState(zStream, stream);
+
+    return result;
+}
+
+extern "C" int32_t Deflate(PAL_ZStream* stream, int32_t flush)
+{
+    z_stream* zStream = GetCurrentZStream(stream);
+
+    int32_t result = deflate(zStream, flush);
+    TransferState(zStream, stream);
+
+    return result;
+}
+
+extern "C" int32_t DeflateEnd(PAL_ZStream* stream)
+{
+    z_stream* zStream = GetCurrentZStream(stream);
+
+    int32_t result = deflateEnd(zStream);
+    
+    End(stream);
+
+    return result;
+}
+
+extern "C" int32_t InflateInit2_(PAL_ZStream* stream, int32_t windowBits)
+{
+    Init(stream);
+
+    z_stream* zStream = GetCurrentZStream(stream);
+
+    int32_t result = inflateInit2(zStream, windowBits);
+    TransferState(zStream, stream);
+
+    return result;
+}
+
+extern "C" int32_t Inflate(PAL_ZStream* stream, int32_t flush)
+{
+    z_stream* zStream = GetCurrentZStream(stream);
+
+    int32_t result = inflate(zStream, flush);
+    TransferState(zStream, stream);
+
+    return result;
+}
+
+extern "C" int32_t InflateEnd(PAL_ZStream* stream)
+{
+    z_stream* zStream = GetCurrentZStream(stream);
+
+    int32_t result = inflateEnd(zStream);
+    
+    End(stream);
+
+    return result;
+}
+
+extern "C" int32_t IsCrc32Available()
+{
+    return true;
+}
```

Comment:
OK - I will change it to return true from that managed side.

I thought that crc32 was only available in recent versions of zlib, and that we might hit a Unix platform that had an older version without crc32. I thought that was why it wasn't available on Windows - because we had compiled an older version. I didn't know it was because we trimmed down the zlib code we compiled for Windows.